include/seqan3/io/record.hpp

```diff
@@ -0,0 +1,312 @@
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+/*!\file
+ * \brief Provides the seqan3::record template and the seqan3::field enum.
+ * \author Hannes Hauswedell <hannes.hauswedell AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <tuple>
+
+#include <meta/meta.hpp>
+
+#include <seqan3/core/platform.hpp>
+
+namespace seqan3
+{
+
+// ----------------------------------------------------------------------------
+// enum field
+// ----------------------------------------------------------------------------
+
+/*!\brief An enumerator for the fields used in file formats.
+ * \ingroup io
+ *
+ * Some of the fields are shared between formats.
+ */
+enum class field
+{
+    SEQ,        //!< The "sequence", usually a range of nucleotides or amino acids.
+    ID,         //!< The identifier, usually a string.
+    QUAL,       //!< The qualities, usually in phred-score notation.
+    SEQ_QUAL,   //!< Sequence and qualities combined in one range.
+};
+
+// ----------------------------------------------------------------------------
+// fields
+// ----------------------------------------------------------------------------
+
+/*!\brief       A class template that holds a choice of seqan3::field.
+ * \ingroup io
+ * \tparam fs   The fields you wish to be present in the seqan3::record returned by your file.
+ * \see         seqan3::record
+ *
+ * This class acts as a compile time list of seqan3::field elements. It used in specialising file classes
+ * to determine the elements in a seqan3::record.
+ *
+ * ### Example
+ *
+ * ```cpp
+ * // specify custom field combination/order to file:
+ * sequence_file_in fin{"/tmp/my.fasta", fields<field::ID, field::SEQ>{}};
+ *
+ * auto record = fin.back(); // get current record, in this case the first
+ *
+ * // record is tuple-like type, but allows access via field identifiers:
+ * auto & id = get<field::ID>(record);
+ * auto & seq = get<field::SEQ>(record);
+ * ```
+ *
+ */
+template <field ...fs>
+struct fields
+{
+    //!\privatesection
```

Comment:
Why only documented as private?

Reply:
because users don't need to call this stuff, they just pass `fields<field::SEQ, field::QUAL>` to their file constructor.